models/repo/archive_download_count.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package repo
+
+import (
+	"context"
+
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/git"
+	api "code.gitea.io/gitea/modules/structs"
+)
+
+// RepoArchiveDownloadCount counts all archive downloads for a tag
+type RepoArchiveDownloadCount struct { //nolint:revive
+	ID     int64           `xorm:"pk autoincr"`
+	RepoID int64           `xorm:"index unique(s)"`
+	Type   git.ArchiveType `xorm:"unique(s)"`
+	Tag    string          `xorm:"index unique(s)"`
+	Count  int64
+}
+
+func init() {
+	db.RegisterModel(new(RepoArchiveDownloadCount))
+}
+
+// CountArchiveDownload adds one download the the given archive
+func CountArchiveDownload(ctx context.Context, repoID int64, tp git.ArchiveType, tag string) error {
+	var counter RepoArchiveDownloadCount
+	has, err := db.GetEngine(ctx).Where("repo_id = ?", repoID).And("`type` = ?", tp).And("tag = ?", tag).Get(&counter)
+	if err != nil {
+		return err
+	}
+
+	if has {
+		// The archive already exists in the database, so let's increase the counter
+		_, err = db.GetEngine(ctx).Incr("count").ID(counter.ID).Update(new(RepoArchiveDownloadCount))
```

Comment:
You can try update directly and then if the return parameter updateed is zero do an insert operation.